### PR TITLE
Use several nice newish Rust features

### PR DESCRIPTION
**Pull request by bradjc**
### Pull Request Overview

As a push towards "Rust 2018", Rust is advertising several new features that are actually quite nice for Tock. This site: https://rust-lang-nursery.github.io/edition-guide/2018/index.html explains them.

What we really benefit from is that before we had to do:

```rust
impl<'a, A: Alarm + 'a> MyStruct<'a, A>
```

Now rust will just figure all that out, and we only need:

```rust
impl<A: Alarm> MyStruct<'a, A>
```

And the simple case is clean now. Before:

```rust
impl<'a> MyOtherStruct<'a>
```

now:

```rust
impl MyOtherStruct<'a>
```

This is I think a lot closer to what someone new to rust would try to write.

### Testing Strategy

This pull request was tested by compiling.

### TODO or Help Wanted
n/a

### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make formatall`.

### CODE REVIEW

**Comment by bradjc** [comment]
Hopefully the docs will build on the new nightly.

File: boards/imix/src/icmp_lowpan_test.rs [hunk 1]
```diff
@@ -147,7 +147,7 @@ impl<'a, A: time::Alarm> capsules::net::icmpv6::icmpv6_send::ICMP6SendClient
     }
 }
 
-impl<'a, A: time::Alarm + 'a> LowpanICMPTest<'a, A> {
+impl<'a, A: time::Alarm> LowpanICMPTest<'a, A> {
```
Comment (brghena) [inline boards/imix/src/icmp_lowpan_test.rs]: Why can't we remove the first `'a` here? (and numerous other places such as ipv6_lowpan_test.rs, sixlowpan_dummy.rs, udp_lowpan_test.rs, and traitobj_list.rs)

Comment (bradjc) [reply boards/imix/src/icmp_lowpan_test.rs]: I think we can. Commits welcome :) (It's a test file so my motivation was low).

Comment (ppannuto) [reply boards/imix/src/icmp_lowpan_test.rs]: First try doesn't "just work":

```
   Compiling imix v0.1.0 (file:///Volumes/code/helena-project/tock/boards/imix)
error[E0106]: missing lifetime specifier
   --> src/icmp_lowpan_test.rs:150:22
    |
150 | impl<A: time::Alarm> LowpanICMPTest<A> {
    |                      ^^^^^^^^^^^^^^^^^ expected lifetime parameter

error[E0106]: missing lifetime specifier
   --> src/icmp_lowpan_test.rs:157:10
    |
157 |     ) -> LowpanICMPTest<A> {
    |          ^^^^^^^^^^^^^^^^^ expected lifetime parameter
    |
    = help: this function's return type contains a borrowed value, but the signature does not say which one of `icmp_sender`'s 2 lifetimes it is borrowed from

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0106`.
```

Messing around a bit I couldn't puzzle out any other thing that would make the compiler happy, but I didn't try too hard

**Comment by bradjc** [comment]
Thoughts?

**Pull request merged**